README.md
<br>

<div align="center">
<img width="456" src="https://raw.githubusercontent.com/wayofdev/ansible-role-tpl/master/assets/logo.gh-light-mode-only.png#gh-light-mode-only">
<img width="456" src="https://raw.githubusercontent.com/wayofdev/ansible-role-tpl/master/assets/logo.gh-dark-mode-only.png#gh-dark-mode-only">
</div>

<br>

<br>

<div align="center">
<a href="https://actions-badge.atrox.dev/wayofdev/ansible-role-mac-cfg/goto"><img alt="Build Status" src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Factions-badge.atrox.dev%2Fwayofdev%2Fansible-role-mac-cfg%2Fbadge&style=flat-square"/></a>
<a href="https://galaxy.ansible.com/wayofdev/maccfg"><img alt="Ansible Role" src="https://img.shields.io/ansible/role/59609?style=flat-square"/></a>
<a href="https://github.com/wayofdev/ansible-role-mac-cfg/tags"><img src="https://img.shields.io/github/v/tag/wayofdev/ansible-role-mac-cfg?sort=semver&style=flat-square" alt="Latest Version"></a>
<a href="https://galaxy.ansible.com/wayofdev/maccfg"><img alt="Ansible Quality Score" src="https://img.shields.io/ansible/quality/59609?style=flat-square"/></a>
<a href="https://galaxy.ansible.com/wayofdev/maccfg"><img alt="Ansible Role" src="https://img.shields.io/ansible/role/d/59609?style=flat-square"/></a>
<a href="LICENSE"><img src="https://img.shields.io/github/license/wayofdev/ansible-role-mac-cfg.svg?style=flat-square&color=blue" alt="Software License"/></a>
<a href="#"><img alt="Commits since latest release" src="https://img.shields.io/github/commits-since/wayofdev/ansible-role-mac-cfg/latest?style=flat-square"></a>
</div>

<br>

# Ansible Role: macOS config Automation

Role is used to automate use of macOS [defaults](https://support.apple.com/en-gb/guide/terminal/apda49a1bb2-577e-4721-8f25-ffc0836f6997/mac) command to configure system settings. Also, additionally, there is available [duti](https://github.com/moretension/duti) tool, which assigns applications to default document types. And all that in automated way!

Example mods that come enabled by default:

* Finder - Disable the warning before emptying the Trash
* Finder - Show bottom path bar by default
* Finder - Disable the warning when changing a file extension
* Text - Do not disable smart quotes, but set them as simple one's, instead of smart
* and many more...

These defaults are used by wayofdev members, but you may override them by defining your own playbook with your config.

If you **like/use** this role, please consider **starring** it. Thanks!

<br>

## 🗂 Table of contents

* [Requirements](#-requirements)
* [Role Variables](#-role-variables)
  * [Applications](#-applications)
  * [Dock](#-dock)
  * [Finder](#-finder)
  * [Input](#-input)
  * [Language & Region](#-language--region)
  * [Network](#-network)
  * [Power](#-power)
  * [Safari](#-safari)
  * [Screensaver](#-screensaver)
  * [Screenshots](#-screenshots)
  * [Security](#-security)
  * [UI](#-ui)
* [Example Playbook](#-example-playbook)
* [Development](#-development)
* [Testing](#-testing)
  * [on localhost](#-on-localhost)
  * [over SSH](#-over-ssh)
* [Dependencies](#-dependencies)
* [Compatibility](#-compatibility)
* [License](#-license)
* [Author Information](#-author-information)
* [Credits and Resources](#-credits-and-resources)
* [Contributors](#-contributors)

<br>

## 📑 Requirements

  - **Homebrew**: Requires `homebrew` already installed (you can use `wayofdev.homebrew` to install it on your macOS).
  - Up-to-date version of ansible. During maintenance/development, we stick to ansible versions and will use new features if they are available (and update `meta/main.yml` for the minimum version).
- Compatible OS. See [compatibility](#-compatibility) table.
- Role has dependencies on third-party roles on different operating systems. See `requirements.yml` and [dependencies](#-dependencies) section.

<br>

## 🔧 Role Variables

Section shows all possible variants of adding, moving, replacing and removing of applications, spacers, folders. Available variables are listed below, along with example values (see `defaults/main.yml`):

<br>

### → Applications

Configures some of built-in macOS applications.

* Activity Monitor

```yaml
# Show the main window when launching app.
maccfg_activity_monitor_show_main_window: true

# Show all processes.
maccfg_activity_monitor_show_category: 0

# Sort Activity Monitor results
maccfg_activity_monitor_sort_by: "CPUUsage"

# Activity Monitor: Sort Direction
# 0 goes for DSC
maccfg_activity_monitor_sort_direction: 0
```

* Mail.app

```yaml
# Copy email addresses as `foo@example.com` instead of `Foo Bar <foo@example.com>` in Mail.app.
maccfg_mail_use_advanced_copy_format: false
```

*  MacOS Photos.app

```yaml
# Prevent Photos from opening automatically when devices are plugged in.
maccfg_photos_disable_hotplug: true
```

* Other applications

```YAML
# Add iOS Simulator to Launchpad.
maccfg_other_create_simulator_symlink: false
```

<br>

### → Dock

```yaml
# Drag a file over an icon in the Dock, hover, and the application will open.
# https://macos-defaults.com/misc/enable-spring-load-actions-on-all-items.html
maccfg_dock_enable_spring_load_everywhere: true

# Don’t show recent applications in Dock
maccfg_dock_show_recents: false
```

* Hot Corners

```yaml
### Hot Corners

# Possible values:
#  0: no-op
#  2: Mission Control
#  3: Show application windows
#  4: Desktop
#  5: Start screen saver
#  6: Disable screen saver
#  7: Dashboard
# 10: Put display to sleep
# 11: Launchpad
# 12: Notification Center

# Bottom right screen corner → Start screen saver
maccfg_dock_bottom_left_corner: 5
maccfg_dock_bottom_left_modifier: 0

# Bottom right screen corner → Launchpad
maccfg_dock_bottom_right_corner: 11
maccfg_dock_bottom_right_modifier: 0
```

* Mission Control

```yaml
# Mission Control - Speed up animations
maccfg_dock_expose_animation_duration: "0.1"
```

<br>

### → Finder

```yaml
# Allow quitting via ⌘ + Q; doing so will also hide desktop icons
maccfg_finder_allow_quitting: true

# What path should new finder windows open
# Computer     : `PfCm`
# Volume       : `PfVo`
# $HOME        : `PfHm`
# Desktop      : `PfDe`
# Documents    : `PfDo`
# Recents      : `PfAF`
# Other…       : `PfLo`
#
# Set Desktop as the default location for new Finder windows by default
maccfg_finder_new_window_target: "PfDe"
maccfg_finder_new_window_target_path: "file://~/Desktop/"

# Set search scope.
# This Mac       : `SCev`
# Current Folder : `SCcf`
# Previous Scope : `SCsp`
maccfg_finder_search_scope: "SCcf"

# Set preferred view mode.
# Icon View   : `icnv`
# List View   : `Nlsv`
# Column View : `clmv`
# Cover Flow  : `Flwv`
# Use column view in all Finder windows by default
maccfg_finder_view_mode: "clmv"

# Show path bar in Finder (Example: 751 items ... 351 GB left ...)
maccfg_finder_show_path_bar: true

# Keep folders on top when sorting by name
maccfg_finder_keep_folders_on_top: true

# Disable the warning when changing a file extension
maccfg_finder_disable_extension_change_warning: true

# Show status bar
maccfg_finder_show_status_bar: true

# Show bottom path bar by default
maccfg_finder_show_bottom_path_bar: true

# Avoid creating .DS_Store files on network volumes
maccfg_finder_avoid_dsstore_on_network_volumes: true

# Avoid creating .DS_Store files on USB volumes
maccfg_finder_avoid_dsstore_on_usb_volumes: true

# Disable the warning before emptying the Trash
maccfg_finder_disable_trash_warning: true

# Display full POSIX path as Finder window title
maccfg_finder_show_full_title: true

# Show all filename extensions
maccfg_finder_show_all_file_exts: false

# Expand the following File Info panes: General, Open With, Sharing & Permissions
maccfg_finder_expand_fileinfo_panels: true

# Show icons for external hard drives on Desktop
maccfg_finder_desktop_show_ext_hard_drives: true

# Show icons for hard drives on Desktop
maccfg_finder_desktop_show_hard_drives: false

# Show icons for mounted servers on Desktop
maccfg_finder_desktop_show_mounted_servers: false

# Show icons for removable media on Desktop
maccfg_finder_desktop_show_removable_media: true

# Show the ~/Library folder in finder
maccfg_finder_unhide_library_folder: true

# Show the ~/Users folder in finder
maccfg_finder_unhide_users_folder: true

# Install duti - utility to allow to change default file types.
maccfg_finder_install_duti: true

# Prefer custom editor over default TextEdit for plain text files.
# Works only if flag 'install_duti' is set to true
# Examples
# com.sublimetext.3
# com.sublimetext.4
# com.microsoft.VSCode
maccfg_finder_default_editor: "com.sublimetext.4"
```

<br>

### → Input

```yaml
# Disable natural scroll for mouse and trackpad.
maccfg_input_use_native_macos_scroll: false
```

* Keyboard

```yaml
# For 0 - Use F1, F2, etc. keys as Brightness/Media
# For 1 - Use F1, F2, etc. keys as standard function keys on external keyboards each key.
maccfg_input_keyboard_fn_key_state: 1

# What should happen when pressing Fn key?
# 0 - Do Nothing
# 1 - Change Input Source
# 2 - Show Emoji & Symbols
# 3 - Start Dictation (Press Twice)
maccfg_input_keyboard_fn_usage_type: 2
```

* Text

```yaml
# Disable text auto-correction in native apps.
maccfg_input_text_disable_auto_correction: true

# Disable text autocorrection in web.
maccfg_input_text_disable_web_auto_correction: true

# Do not disable smart quotes, but set them as simple one's, instead of smart.
# Use "" for double quotes and '' for single quotes
maccfg_input_text_use_simple_quotes: true
```

* Mouse

```yaml
# Magic mouse should run in two button mode
maccfg_input_mouse_two_button_mode: true

# Magic mouse - enable double tap gesture for zoom
maccfg_input_mouse_double_tap_gesture: true
```

* Trackpad

```yaml
# Map bottom right corner to right-click for integrated trackpad
maccfg_input_trackpad_bottom_right_click_trackpad: true

# Map bottom right corner to right-click for bluetooth trackpad
maccfg_input_trackpad_bottom_right_click_bt_trackpad: true
```

* Shortcuts

```yaml
# Allow global app shortcuts
maccfg_input_shortcuts_allow: true

# Enable global shortcut – ⌥⌘, brings up System Preferences dialog (opt+cmd+comma)
maccfg_input_shortcuts_opt_cmd_comma: true
```

<br>

### → Language & Region

```yaml
# Set the timezone; see `sudo systemsetup -listtimezones` for other values
maccfg_lang_timezone: "Europe/Riga"

# Set main languages - English US, Latvian - LV, Latvian - RU.
maccfg_lang_languages:
  - en-US
  - lv-LV
  - ru-LV

maccfg_lang_locale: "en_US@currency=EUR"

# Use metric measurement units.
maccfg_lang_metric_units: true

# Set "AppleMeasurementUnits" setting to centimeters.
maccfg_lang_apple_units: Centimeters

# Enables the input menu in the menu bar.
maccfg_lang_inputs_in_status_bar: true

# List of input sources.
maccfg_lang_input_sources:
  # Default input sources.
  - name: "ABC"
    id: 252
  - name: "Latvian"
    id: 30765
  - name: "Russian - Phonetic"
    id: 19457
```

<br>

### → Network

```yaml
# Perform hostname change? On remote machines can cause connection lose.
maccfg_network_change_hostname: true

# This mac-book's hostname
# On macOS, this module uses scutil to set HostName, ComputerName, and LocalHostName
maccfg_network_hostname: "mbpro-{{ remote_regular_user }}"
```

<br>

### → Power

Section controls how mac power will be managed on charger or battery:

```yaml
# Mostly here 'pmset' utility is used. For reference see:
# https://www.dssw.co.uk/reference/pmset.html
#
# To get total regular & safe sleep set following parameters to these values:
# powernap: false
# womp: false
# tcpkeepalive: false
# ttyskeepawake false
#
# Restart automatically if the computer freezes
maccfg_power_restart_on_freeze: true

###
# Settings to apply when device is on A/C charger.
###

# Wake the machine when the laptop lid is opened
maccfg_power_on_charger_lidwake: true

# Prevent idle system sleep when any tty (e.g. remote login session) is ’active’.
maccfg_power_on_charger_ttyskeepawake: true

# https://support.apple.com/en-gb/guide/mac-help/mh40773/mac
#
# Power Nap, available on Mac computers with flash memory, lets some Mac computers stay up to date
# even while they’re sleeping. When your Mac goes to sleep, Power Nap activates periodically to update information.
# The information that’s updated depends on whether your Mac is running on battery power (a Mac notebook computer)
# or plugged into a power adapter (a Mac notebook computer or Mac desktop computer).
# When your Mac is asleep and using battery power, Power Nap:
# * Checks for new messages in Mail
# * Updates events in Calendar
# * Updates other iCloud events
# When your Mac is plugged into a power adapter, Power Nap can also do activities such as
# downloading software updates and performing Time Machine backups.
maccfg_power_on_charger_powernap: true

# Set hibernate mode:
# hibernatemode = 0 (binary 0000) by default on supported desktops. The system will not back memory up to
#   persistent storage. The system must wake from the contents of memory; the system will lose context on
#   power loss. This is, historically, plain old sleep.
#
# hibernatemode = 3 (binary 0011) by default on supported portables. The system will store a copy of mem-ory memory
#   ory to persistent storage (the disk), and will power memory during sleep. The system will wake from
#   memory, unless a power loss forces it to restore from disk image.
#
# hibernatemode = 25 (binary 0001 1001) is only settable via pmset. The system will store a copy of mem-ory memory
#   ory to persistent storage (the disk), and will remove power to memory. The system will restore from
#   disk image. If you want "hibernation" - slower sleeps, slower wakes, and better battery life, you
#   should use this setting.
maccfg_power_on_charger_hibernatemode: 3

# Display sleep timer
maccfg_power_on_charger_displaysleep: 15

# Wake on ethernet magic packet. Same as "Wake for network access" in the Energy Saver preferences.
maccfg_power_on_charger_womp: true

# System sleep timer (value in minutes, or 0 to disable)
maccfg_power_on_charger_sleep: 0

# Wake for network access
maccfg_power_on_charger_tcpkeepalive: true

# Display sleep will use an intermediate half-brightness state between full brightness and fully off
maccfg_power_on_charger_halfdim: true

# Wake the machine when power source (AC/battery) is changed
maccfg_power_on_charger_acwake: false

# https://apple.stackexchange.com/a/342286
# Not used with SSD's
maccfg_power_on_charger_disksleep: 0

###
# Settings to apply when device is using battery.
###
maccfg_power_on_battery_lidwake: true
maccfg_power_on_battery_ttyskeepawake: false
maccfg_power_on_battery_powernap: false
maccfg_power_on_battery_hibernatemode: 3
maccfg_power_on_battery_displaysleep: 15
maccfg_power_on_battery_sleep: 30
maccfg_power_on_battery_tcpkeepalive: false
maccfg_power_on_battery_halfdim: true
maccfg_power_on_battery_acwake: false
maccfg_power_on_battery_disksleep: 0
```

<br>

### → Safari

```yaml
# Safari: Enable Develop menu and the Web Inspector in Safari 14+
maccfg_safari_web_inspector: true

# Safari: Change default homepage
maccfg_safari_homepage: "https://app.daily.dev/"

# Safari: Setup new tab behavior
# 0: Homepage
# 1: Empty Page
# 2: Same Page
# 3: Bookmarks
# 4: Top Sites
maccfg_safari_tab_behaviour: 0

# Safari: Setup new window behavior
# 0: Homepage
# 1: Empty Page
# 2: Same Page
# 3: Bookmarks
# 4: Top Sites
maccfg_safari_window_behaviour: 0
```

<br>

### → Screensaver

```yaml
# Set default system screensaver
# Available:
# "Album Artwork"
# "Arabesque"
# "Drift"
# "FloatingMessage"
# "Flurry"
# "Hello"
# "Monterey"
# "Random"
# "Shell"
# "Word of the Day"
maccfg_screensaver_name: "Flurry"

# Show clock when screensaver is running
maccfg_screensaver_show_clock: true

# Show screen saver after:
# Please choose these values, as OSX use dropdown in UI
#
# 0 - do not use screensaver
# 60 - 1 minute
# 120 - 2 minutes
# 300 - 5 minutes
# 600 - 10 minutes
# 1200 - 20 minutes
# 1800 - 30 minutes
# 3600 - 1 hour
maccfg_screensaver_start_after: 300
```

<br>

### → Screenshots

```yaml
# Save to custom folder instead of Desktop.
maccfg_screenshots_directory: "/Users/{{ remote_regular_user }}/Screenshots"

# Change the default file name prefix.
maccfg_screenshots_use_prefix: "src"
```

<br>

### → Security

Section of  macOS's **System Preferences** → **Security & Privacy** and **Software Update**.

```yaml
# Require password when screen awakes.
maccfg_security_require_pass_after_awake: true

# Require password immediately after sleep or screen saver begins
# 0 is for immediate password require
maccfg_security_require_pass_delay: 0

# Turn off Feedback Assistant data auto-gathering.
maccfg_privacy_apple_feedback_assistant: false

# Turn off personalized ads, a little less tracking from apple side...
maccfg_privacy_disable_personalized_ads: true
```

* **Firewall**

```yaml
# Turn on system firewall.
# Turned off by default.
# If this role is used with dev-env-osx playbook, then lulu firewall is used instead.
# Modes:
# https://discussions.apple.com/thread/3148672
# 0 - disabled
# 1 - enabled in "open" mode
# 2 - enabled in "strict" mode
maccfg_firewall_mode: 0
```

* **Updates**

```yaml
# Check for software updates once per week
# Default: 7 (in days)
maccfg_updates_delay_days: 7
```

* **Filevault**

```yaml
# FileVault full-disk encryption with a 256-bit key to help prevent unauthorized access to the information on startup disk.
# Can be risky to enable it on remote machines!
maccfg_filevault_enabled: true
```

<br>

### → UI

```yaml
# Expand save panel by default.
maccfg_ui_expand_save_panel: true

# Expand save panel by default.
maccfg_ui_expand_print_panel: true

# Automatically quit printer app once the print jobs complete.
maccfg_ui_quit_print_job: true

# Reveal IP address, hostname, OS version, etc. when clicking the clock in the login window (only).
maccfg_ui_login_clock_clickable: true

# Show bluetooth in menu bar.
maccfg_ui_menu_bar_bluetooth_icon: true
```

<br>

## 📗 Example Playbook

```yaml
---
- hosts: all

  # is needed when running over SSH
  environment:
    - PATH: "/usr/local/bin:/usr/local/sbin:/opt/homebrew/bin:{{ ansible_env.PATH }}"

  vars:
    maccfg_filevault_enabled: false

  roles:
    - wayofdev.homebrew
    - wayofdev.maccfg
```

<br>

## ⚙️ Development

To install dependencies and start development you can check contents of our `Makefile`

**Install** [poetry](https://github.com/python-poetry/poetry) using [poetry-bin](https://github.com/gi0baro/poetry-bin) and all dev python dependencies:

```bash
$ make install
```

**Install** only python dependencies, assuming that you already have poetry:

```bash
$ make install-deps
```

**Install** all git hooks:

```bash
$ make hooks
```

**Lint** all role files:

```bash
$ make lint
```

<br>

## 🧪 Testing

You can check `Makefile` to get full list of commands for remote and local testing. For local testing you can use these comands to test whole role or separate tasks:

### → on localhost

> :warning: **Notice**: By defaut all tests are running against your local machine!

```bash
# run molecule tests on localhost
$ poetry run molecule test --scenario-name default-macos-on-localhost -- -vvv

# or with make command
$ make m-local

# choose which tags will be included
$ export TASK_TAGS="dock-validate,dock-install"; make m-local
```

<br>

### → over SSH

```bash
# run molecule scenarios against remote machines over SSH
# this will need VM setup and configuration
$ poetry run molecule test --scenario-name default-macos-over-ssh -- -vvv

$ make m-remote

# tags also can be passed
$ export TASK_TAGS="dock-validate,dock-install"
$ make m-remote
```

<br>

## 📦 Dependencies

Installation handled by `Makefile` and requirements are defined in `requirements.yml`

  - [wayofdev.homebrew](https://github.com/wayofdev/ansible-role-homebrew) - soft dependency, required if Homebrew isn't installed yet and only if you need `duti`
  - [ansible.community.general](https://docs.ansible.com/ansible/latest/collections/community/general/index.html)

<br>

## 🧩 Compatibility

This role has been tested on these systems:

| system / container | tag      |
| :----------------- | -------- |
| macos              | monterey |
| macos              | big-sur  |

<br>

## 🤝 License

[![Licence](https://img.shields.io/github/license/wayofdev/ansible-role-dock?style=for-the-badge&color=blue)](./LICENSE)

<br>

## 🙆🏼‍♂️ Author Information

This role was created in **2022** by [lotyp / wayofdev](https://github.com/wayofdev).

<br>

## 🧱 Credits and Resources

**Useful resources:**

* [duti](https://github.com/moretension/duti)

<br>

## 🫡 Contributors

<img align="left" src="https://img.shields.io/github/contributors-anon/wayofdev/ansible-role-dock?style=for-the-badge"/>

<a href="https://github.com/wayofdev/ansible-role-maccfg/graphs/contributors">
  <img src="https://opencollective.com/wod/contributors.svg?width=890&button=false">
</a>

<br>


CHANGELOG.md
# Changelog

## [1.1.2](https://github.com/wayofdev/ansible-role-mac-cfg/compare/v1.1.1...v1.1.2) (2022-07-04)


### Bug Fixes

* fatal error when trying to add global shortcuts ([698a2d9](https://github.com/wayofdev/ansible-role-mac-cfg/commit/698a2d9d1ce8317d23328fa44f599f802b7cc3fb))

## [1.1.1](https://github.com/wayofdev/ansible-role-mac-cfg/compare/v1.1.0...v1.1.1) (2022-07-04)


### Bug Fixes

* minor settings tune ([0516b09](https://github.com/wayofdev/ansible-role-mac-cfg/commit/0516b09fc1537e8bbeaca65cbcb8bf0e1b510adc))
* typo in variable ([bb884f5](https://github.com/wayofdev/ansible-role-mac-cfg/commit/bb884f5c7d4d215c650617aa2dde7b5bc80ec6ed))

## [1.1.0](https://github.com/wayofdev/ansible-role-mac-cfg/compare/v1.0.0...v1.1.0) (2022-07-02)


### Miscellaneous Chores

* release 1.1.0 ([84d8b57](https://github.com/wayofdev/ansible-role-mac-cfg/commit/84d8b572e2d48697304041a1fd745ce82efdbcee))

## 1.0.0 (2022-06-30)


### ⚠ BREAKING CHANGES

* refactoring code to comply best practices and latest wod template

### Features

* adding firevault support, menubar bluetooth ([ce5fcf4](https://github.com/wayofdev/ansible-role-mac-cfg/commit/ce5fcf47fd7d89e040474c060edbd3b0080ca018))
* adding variables ([cb8e1b5](https://github.com/wayofdev/ansible-role-mac-cfg/commit/cb8e1b5fa9816fa8d546c6003df8410350f58b13))
* disable personalized ads ([475ab5b](https://github.com/wayofdev/ansible-role-mac-cfg/commit/475ab5bf3c8e3bdddc224e754e5846b3da7d0893))
* initial repository commit ([00c2704](https://github.com/wayofdev/ansible-role-mac-cfg/commit/00c2704d7ae6d15c4545381986bde113fa845241))
* moving static values to variables ([12894f5](https://github.com/wayofdev/ansible-role-mac-cfg/commit/12894f537d5b675d7101eea6c32f9292b0081e2d))
* new settings ([64e6570](https://github.com/wayofdev/ansible-role-mac-cfg/commit/64e657090ee812ec4ababf53858c3bd59467a74e))
* refactoring code to comply best practices and latest wod template ([5fb1f0b](https://github.com/wayofdev/ansible-role-mac-cfg/commit/5fb1f0b16d3bc3bee05d58b805cca7faed9fa3ff))


### Bug Fixes

* bool/int type mismatch ([2c81803](https://github.com/wayofdev/ansible-role-mac-cfg/commit/2c81803a17c15bb3a54483608938ef1a0326f0fa))
* bool/int type mismatch ([e17d0a9](https://github.com/wayofdev/ansible-role-mac-cfg/commit/e17d0a931888eb06b8db6b4ea269969b790ceca7))
* escaping quotes ([1a6e321](https://github.com/wayofdev/ansible-role-mac-cfg/commit/1a6e321627f7c379b56da45950bf4966df0d47ac))
* escaping quotes ([2eaa655](https://github.com/wayofdev/ansible-role-mac-cfg/commit/2eaa6554b65c8a768fc65e53c5b457ba4e13531f))
* github runners arent laptops ([189fce1](https://github.com/wayofdev/ansible-role-mac-cfg/commit/189fce1107c4ee96ff45ef2f53379009e1f282f4))
* github runners does not have battery ([fc3227f](https://github.com/wayofdev/ansible-role-mac-cfg/commit/fc3227fdaa8aa7e5d4cc2eed6c5461e70b6c01c5))
* hostname change causes remote hosts to hang up ([b9ade59](https://github.com/wayofdev/ansible-role-mac-cfg/commit/b9ade59f78d86a607647e971440a5d7e8c8cd56f))
* idempotency failure ([3e26f81](https://github.com/wayofdev/ansible-role-mac-cfg/commit/3e26f819aa2bc40fbee57464df9b9e015b606f1e))
* moving hostname change to variables ([9e17950](https://github.com/wayofdev/ansible-role-mac-cfg/commit/9e1795037b7217f4913d7fd92aaeb6d4c655fb14))
* moving hostname change to variables ([ed7577b](https://github.com/wayofdev/ansible-role-mac-cfg/commit/ed7577be300b9669de4a61d424294bb2e849745d))
* non existent app symlink ([4c35ded](https://github.com/wayofdev/ansible-role-mac-cfg/commit/4c35ded7d90d151f7058c40c79734d0557dcfb65))
* root user does not have FinderInfo attr on Library ([8e0cffe](https://github.com/wayofdev/ansible-role-mac-cfg/commit/8e0cffe44b3b30b23bf5b754a34435e71ec974c0))
* type mismatch fix ([6f3e992](https://github.com/wayofdev/ansible-role-mac-cfg/commit/6f3e992bec1fd639429e7e1afd6d8d6e08b171c1))
* typo in new hostname variables ([23e7320](https://github.com/wayofdev/ansible-role-mac-cfg/commit/23e7320885221dfe6fb25f42e1ed81cabb4e7fb1))
* typo in shell command ([7fa4739](https://github.com/wayofdev/ansible-role-mac-cfg/commit/7fa47392315350babb32073442a2e6dc61be4107))
* typo un duti application ([5a9774a](https://github.com/wayofdev/ansible-role-mac-cfg/commit/5a9774a9a6f37688f317d2070fe982fb9f4ae596))
* typo un duti application ([3c5b46b](https://github.com/wayofdev/ansible-role-mac-cfg/commit/3c5b46b6f3a63c25cfedb62ffb24ad08200695a2))
* unclosed braces in greps ([15f0cc8](https://github.com/wayofdev/ansible-role-mac-cfg/commit/15f0cc894f802ca825efc278069616a24659c765))
* unclosed braces in greps ([3cdcbf1](https://github.com/wayofdev/ansible-role-mac-cfg/commit/3cdcbf18e3e0a4d768f55f2e3fc2ea3f459f4a3e))
* using unicode chars for simple single and double quotes ([b31b767](https://github.com/wayofdev/ansible-role-mac-cfg/commit/b31b767f858bd8cef0903e10827f601f70b642a3))


library/pmset.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# (c) 2022, lotyp <lotyp7.woop@gmail.com>

from ansible.module_utils.basic import AnsibleModule


def parse_pmset_output(output):
    # Parses `pmset -g custom` into a 2-level dict.
    section_name = ''
    result = {}
    for line in output.splitlines():
        if line == '' or 'Sleep On Power Button' in line:
            continue
        if line[0] != ' ' and line[-1] == ':':
            section_name = line[:-1]
            result[section_name] = {}
        else:
            k, v = line.split()
            result[section_name][k] = v
    return result


def run_module():
    module_args = dict(
        on_battery=dict(type='dict', required=False, default=dict()),
        on_charger=dict(type='dict', required=False, default=dict()),
    )
    result = dict(changed=False, diff=dict(before='', after=''))
    module = AnsibleModule(argument_spec=module_args, supports_check_mode=True)

    _, stdout, _ = module.run_command(["pmset", "-g", "custom"], check_rc=True)

    commands = []
    output = parse_pmset_output(stdout)

    def add_diff(block, param, old_value, new_value):
        result['diff']['before'] += '{block}.{param}={val}\n'.format(
            block=block, param=param, val=old_value)
        result['diff']['after'] += '{block}.{param}={val}\n'.format(
            block=block, param=param, val=new_value)

    blocks = [('on_charger', '-c', output['AC Power']), ]

    if 'Battery Power' in output:
        blocks.append(('on_battery', '-b', output['Battery Power']))

    for block, mode_flag, current_values in blocks:
        # Iterate over user-specified parameters and check if there's anything
        # to change.
        for param, value in module.params[block].items():
            if value is None: continue
            if param not in current_values:
                continue
            orig = current_values[param]
            if isinstance(value, int):
                orig = int(orig)
            if orig != value:
                add_diff(block, param, orig, value)
                commands.append(['pmset', mode_flag, param, value])

    if commands:
        result['changed'] = True
    if module.check_mode:
        module.exit_json(**result)

    for command in commands:
        # Explicitly convert all arguments to strings, so that
        # module.run_command doesn't choke on ints.
        cmd = [str(v) for v in command]
        module.run_command(cmd, check_rc=True)

    module.exit_json(**result)


def main():
    run_module()


if __name__ == '__main__':
    main()


library/firevault.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# (c) 2017, Richard Oliver <source@richard-oliver.co.uk>
# (c) 2022, lotyp <lotyp7.woop@gmail.com>

# Known bugs:
# - enabled: True will return a false-success if FileVault is in the
#   process of decrypting the volume.

import subprocess as sp
from ansible.module_utils.basic import AnsibleModule


# for more info about exit codes: man fdesetup
fdesetup_errors = {
    0: "No error, or successful operation.",
    1: "FileVault is Off.",
    2: "FileVault appears to be On but Busy.",
    11: "Authentication error.",
    12: "Parameter error.",
    13: "Unknown command error.",
    14: "Bad command error.",
    15: "Bad input error.",
    16: "Legacy FileVault error.",
    17: "Added users failed error.",
    18: "Unexpected keychain found error.",
    19: "Keychain error. This usually means the FileVaultMaster keychain could not be moved or replaced.",
    20: "Deferred configuration setup missing or error.",
    21: "Enable failed (Keychain) error.",
    22: "Enable failed (CoreStorage) error.",
    23: "Enable failed (DiskManager) error.",
    24: "Already enabled error.",
    25: "Unable to remove user or disable FileVault.",
    26: "Unable to change recovery key.",
    27: "Unable to remove recovery key.",
    28: "FileVault is either off, busy, or the volume is locked.",
    29: "Did not find FileVault information at the specified location.",
    30: "Unable to add user to FileVault because user record could not be found.",
    31: "Unable to enable FileVault due to management settings.",
    32: "FileVault is already active.",
    33: "Command option is unsupported on this file system.",
    34: "An option or parameter is not supported for APFS volumes.",
    35: "An error occurred during FileVault disablement.",
    36: "This computer does not support enabling FileVault.",
    37: "One or more users have a blank password. FileVault cannot be enabled.",
    99: "Internal error."
}


def main():
    active = False

    module = AnsibleModule(
        argument_spec=dict(
            enabled=dict(required=True, type='bool'),
        ),
        supports_check_mode=True
    )

    # Check whether FileVault is enabled
    try:
        sp.check_output(["fdesetup", "isactive"])
        active = True
    except sp.CalledProcessError as e:
        if e.returncode == 1:
            active = False
        else:
            module.fail_json(msg="fdesetup failed: " + fdesetup_errors[e.returncode])

    enabled = module.params.get('enabled')

    if active:
        # FileVault is currently enabled
        if module.check_mode:
            module.exit_json(changed=not enabled)

        if enabled:
            module.exit_json(changed=False);
        else:
            try:
                sp.check_output(["fdesetup", "disable"])
                module.exit_json(changed=True)
            except sp.CalledProcessError as e:
                module.fail_json(msg="fdesetup failed: " + fdesetup_errors[e.returncode])
    else:
        # FileVault is currently disabled
        if module.check_mode:
            module.exit_json(changed=enabled)

        if not enabled:
            module.exit_json(changed=False);
        else:
            try:
                sp.check_output(["fdesetup", "enable", "-defer", "/dev/null", "-showrecoverykey", "-forceatlogin", "0",
                                 "-dontaskatlogout"])
                module.exit_json(changed=True)
            except sp.CalledProcessError as e:
                module.fail_json(msg="fdesetup failed: " + fdesetup_errors[e.returncode])


if __name__ == '__main__':
    main()
